packages/edge-preview-authenticated-proxy/src/index.ts

```diff
@@ -0,0 +1,328 @@
+		throw new RawHttpFailed();
+	}
+	const workerUrl = new URL(url);
+	const remoteUrl = new URL(remote);
+	workerUrl.hostname = remoteUrl.hostname;
+	workerUrl.protocol = remoteUrl.protocol;
+
+	const workerResponse = await fetch(workerUrl, {
+		...request,
+		headers: {
+			...request.headers,
+			"cf-workers-preview-token": token,
+		},
+	});
+	// The client needs the raw headers from the worker
+	// Prefix them with `cf-ew-raw-`, so that response headers from _this_ worker don't interfere
+	const rawHeaders: Record<string, string> = {};
+	for (const header of workerResponse.headers.entries()) {
+		rawHeaders[`cf-ew-raw-${header[0]}`] = header[1];
+	}
+	return new Response(workerResponse.body, {
+		...workerResponse,
+		headers: {
+			...rawHeaders,
+			"Access-Control-Allow-Origin": request.headers.get("Origin") ?? "",
+			"Access-Control-Allow-Method": "*",
+			"Access-Control-Allow-Credentials": "true",
+			"Access-Control-Allow-Headers": "x-cf-token,x-cf-remote",
+			"cf-ew-status": workerResponse.status.toString(),
+			"Access-Control-Expose-Headers": "*",
+			Vary: "Origin",
+		},
+	});
+}
+
+async function updatePreviewToken(url: URL, ctx: ExecutionContext) {
+	const token = url.searchParams.get("token");
+	const prewarmUrl = url.searchParams.get("prewarm");
+	const remote = url.searchParams.get("remote");
+	if (!token || !prewarmUrl || !remote) {
+		throw new TokenUpdateFailed();
+	}
+	if (prewarmUrl && token) {
+		ctx.waitUntil(
+			fetch(prewarmUrl, {
+				method: "POST",
+				headers: {
+					"cf-workers-preview-token": token,
+				},
+			})
+		);
+	}
+	return new Response(null, {
+		status: 307,
+		headers: {
+			Location: url.searchParams.get("suffix") ?? "/",
+			"Set-Cookie": cookie.serialize(
+				"token",
+				JSON.stringify({ token, remote }),
+				{
+					secure: true,
+					sameSite: "none",
+					httpOnly: true,
+					domain: url.hostname,
+				}
+			),
+		},
+	});
+}
+
+async function handleTokenExchange(url: URL) {
+	const exchangeUrl = url.searchParams.get("exchange_url");
+	if (!exchangeUrl) {
+		throw new NoExchangeUrl();
+	}
+	const exchangeRes = await fetch(exchangeUrl);
+	if (exchangeRes.status !== 200) {
+		const exchange = new URL(exchangeUrl);
+		// Clear sensitive token
+		exchange.search = "";
+
+		throw new ExchangeFailed(
+			exchange.href,
+			exchangeRes.status,
+			await exchangeRes.text()
+		);
+	}
+	const session = await exchangeRes.json<{
+		prewarm: string;
+		token: string;
+	}>();
+	if (
+		typeof session.token !== "string" ||
+		typeof session.prewarm !== "string"
+	) {
+		const exchange = new URL(exchangeUrl);
+		// Clear sensitive token
+		exchange.search = "";
+		throw new ExchangeFailed(
+			exchange.href,
+			exchangeRes.status,
+			JSON.stringify(session)
+		);
+	}
+	return Response.json(session, {
+		headers: {
+			"Access-Control-Allow-Origin": "*",
+			"Access-Control-Allow-Method": "POST",
+		},
+	});
+}
+
+async function handleRequest(request: Request, ctx: ExecutionContext) {
```

Comment:
NIT (personal taste)

I would put this function at the top of the file and then the ones that handle each scenario below. This makes reading the file from top to bottom more straightforward.
Also, I would put these JSDOC comments on the definition of the functions themselves rather than where they are called.

As it stands I scroll down and see all these functions at the top that have no context - I don't know how/where they are called and they don't have any docs. And I have have to search down to find this function to work out what is going on.